fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -655,12 +645,26 @@ struct EncryptedRangeFileWriter : public IRangeFileWriter {
 		    AES_256_IV_LENGTH,
 		    getEncryptAuthTokenMode(EncryptAuthTokenMode::ENCRYPT_HEADER_AUTH_TOKEN_MODE_SINGLE),
 		    BlobCipherMetrics::BACKUP);
-		Arena arena;
 		int64_t payloadSize = self->wPtr - self->dataPayloadStart;
-		auto encryptedData = encryptor.encrypt(self->dataPayloadStart, payloadSize, self->encryptHeader, arena);
+		StringRef encryptedData;
+		if (self->options.configurableEncryptionEnabled) {
+			BlobCipherEncryptHeaderRef headerRef;
+			encryptedData = encryptor.encrypt(self->dataPayloadStart, payloadSize, &headerRef, *self->arena);
+			Standalone<StringRef> serialized = BlobCipherEncryptHeaderRef::toStringRef(headerRef);
+			self->arena->dependsOn(serialized.arena());
+			ASSERT(serialized.size() == self->encryptHeader.size());
+			std::memcpy(mutateString(self->encryptHeader), serialized.begin(), self->encryptHeader.size());
+		} else {
+			BlobCipherEncryptHeader header;
+			encryptedData =
+			    encryptor.encrypt(self->dataPayloadStart, payloadSize, &header, *self->arena)->toStringRef();
+			StringRef encryptHeaderStringRef = BlobCipherEncryptHeader::toStringRef(header, *self->arena);
+			ASSERT(encryptHeaderStringRef.size() == self->encryptHeader.size());
+			std::memcpy(mutateString(self->encryptHeader), encryptHeaderStringRef.begin(), self->encryptHeader.size());
```

Comment:
Seems like are doing a memcpy for both payload and encryptionHeader; is there a way to avoid this memcpy operation? 

I think it would be difficult to avoid encryptionHeader memcpy as header needs to be placed at a certain location within the block that needs to be written out, but, the bigger gain might be from avoiding memcpy for the payload. In current implementation, thought `encrypt` API allows caller to supply `arena` to allocate from, the code just doesn't use the feature by supplying a local Arena variable, which isn't tied to the buffer allocation that gets written out.. 

Am not sure how easy it would be to solve though; though we know the encrypted buffer final size (same as plaintext), but the algorithms needs to add a `slack` space == AES_BLOCK_SIZE for the Aes calls to work...

Reply:
We need a way to then maybe pass a StringRef to the encrypt method to use that to allocate the header. This is because there a specific point in the block the payload needs to go. Just passing an arena wouldn't work. In any case i think this is best left for a patch if needed. I don't know whether the effort to do this and get it correctness clean is worth it

Reply:
I think updating `encrypt` method to use client supplied memory is real easy, only catch is it needs to be sufficient to hold `sizeof(plaintext) + AES_BLOCK_SIZE` for the scheme to work.

Reply:
Discussed offline with Nim and we can do the following:
1. For `encrypted` backups, the total block allocation size would be : blockSize + AES_BLOCK_SIZE
2. Update `BlobCipher` to expose an API that allows caller to pass data-ptr to copy encrypted buffer.
3. Update the BackupAgent code to use `file flush` check to account for extra AES_BLOCK_SIZE allocated for encryption assistance. 

Plan is to tackle it as a separate PR.

Reply:
Discussed offline, this will work but will be a TODO